src/ai/defaultai.cc

```diff
@@ -4996,13 +4997,13 @@ BuildingNecessity DefaultAI::check_building_necessity(BuildingObserver& bo,
 		if (!basic_economy_established) {
 			return BuildingNecessity::kForbidden;
 		}
-		const uint16_t min_roads_count =
-		   40 + std::abs(management_data.get_military_number_at(33)) / 2;
+		const int32_t min_roads_count = 40 + std::abs(management_data.get_military_number_at(33)) / 2;
 		if (static_cast<int>(roads.size()) < min_roads_count * (1 + bo.total_count())) {
 			return BuildingNecessity::kForbidden;
 		}
-		bo.primary_priority += (roads.size() - min_roads_count * (1 + bo.total_count())) *
-		                       (2 + std::abs(management_data.get_military_number_at(143)) / 5);
+		bo.primary_priority +=
+		   (static_cast<int>(roads.size()) - min_roads_count * (1 + bo.total_count())) *
```

Comment:
@hessenfarmer 'primary_priority' is declared as int32_t on L496 of ai_help_structs.h. So I think the error was because (if I interpreted things correctly) one of the operands is a unsigned long (roads.size()). So trying to multiply an int with an unsigned long could overflow the int32_t target.

Codeql docs have a bit of into about that: [Multiplication result converted to larger type](https://codeql.github.com/codeql-query-help/cpp/cpp-integer-multiplication-cast-to-long/)

Though like you I'm not sure if that's the best option. Maybe I'm wrong, but I think that 'roads.size()'  is ever  '>' than INT_MAX, casting would skew the value and so the multiplication result would be very wrong.

Reply:
changing the declaration of primary_priority would not help as the possible error happens before the value is assigned to it.    
the problem is that an integer multiplication results always in the data type of the biggest operand. The conversion (implicit cast to the target type is done in the asignment step after the multiplication. 
if code ql finds an instance where some narrow integers are multiplied and the result is a wider type that this is intentional to prevent overflow which it doesn't. this is explained in the example help text you provided. so we need to upcast one operand to a wider type to prevent such overflow.

Reply:
I am stillnot convinced by this cast as it is effectively downcasting the value of roads size. furthermore it seems the bracckets are wrong changing the result of the calculation.